src/audio/fork_trigger.c

```diff
@@ -0,0 +1,280 @@
+ * COMP_STATE_PREPARE		-> [COMP_TRIGGER_RESET]		-> COMP_STATE_READY
+ * COMP_STATE_PREPARE		-> [COMP_TRIGGER_PREPARE]	-> COMP_STATE_PREPARE
+ * COMP_STATE_PAUSED		-> [COMP_TRIGGER_PRE_RELEASE]	-> COMP_STATE_PRE_ACTIVE
+ * COMP_STATE_PAUSED		-> [COMP_TRIGGER_STOP]		-> COMP_STATE_PREPARE
+ * COMP_STATE_PAUSED		-> [COMP_TRIGGER_XRUN]		-> COMP_STATE_PAUSED
+ * COMP_STATE_PAUSED		-> [COMP_TRIGGER_RESET]		-> COMP_STATE_READY
+ * COMP_STATE_ACTIVE		-> [COMP_TRIGGER_STOP]		-> COMP_STATE_PREPARE
+ * COMP_STATE_ACTIVE		-> [COMP_TRIGGER_XRUN]		-> COMP_STATE_ACTIVE
+ * COMP_STATE_ACTIVE		-> [COMP_TRIGGER_PAUSE]		-> COMP_STATE_PAUSED
+ * COMP_STATE_ACTIVE		-> [COMP_TRIGGER_RESET]		-> COMP_STATE_READY
+ * COMP_STATE_PRE_ACTIVE	-> [COMP_TRIGGER_START]		-> COMP_STATE_ACTIVE
+ * COMP_STATE_PRE_ACTIVE	-> [COMP_TRIGGER_RELEASE]	-> COMP_STATE_ACTIVE
+ *
+ * Target states
+ * COMP_STATE_INIT		<- *init*
+ * COMP_STATE_READY		<- [COMP_TRIGGER_RESET]		# PCM_FREE -> pipeline_reset()
+ * COMP_STATE_PREPARE		<- [COMP_TRIGGER_STOP]
+ * COMP_STATE_PREPARE		<- [COMP_TRIGGER_XRUN]
+ * COMP_STATE_PREPARE		<- [COMP_TRIGGER_PREPARE]	# PCM_PARAMS -> pipeline_prepare()
+ * <unused>			<- [COMP_TRIGGER_SUSPEND]
+ */
+
+static int source_status_count(struct comp_dev *dev, unsigned int status)
+{
+	struct comp_buffer *source;
+	struct list_item *blist;
+	int count = 0;
+
+	/* count sources with state == status */
+	list_for_item(blist, &dev->bsource_list) {
+		source = container_of(blist, struct comp_buffer, sink_list);
+		comp_info(source->source, "source state %u", source->source->state);
+		if (source->source->state == status)
+			count++;
+	}
+
+	return count;
+}
+
+static int sink_status_count(struct comp_dev *dev, unsigned int status)
+{
+	struct comp_buffer *sink;
+	struct list_item *blist;
+	int count = 0;
+
+	/* count sinks with state == status */
+	list_for_item(blist, &dev->bsink_list) {
+		sink = container_of(blist, struct comp_buffer, source_list);
+		comp_info(sink->sink, "sink state %u", sink->sink->state);
+		if (sink->sink->state == status)
+			count++;
+	}
+
+	return count;
+}
+
+static unsigned int fork_status_count(struct comp_dev *dev,
+				      unsigned int status)
+{
+	switch (dev_comp_type(dev)) {
+	case SOF_COMP_MIXER:
+	case SOF_COMP_MUX:
+		return source_status_count(dev, status);
+	case SOF_COMP_DEMUX:
+		return sink_status_count(dev, status);
+	default:
+		return 0;
+	}
+}
+
+static int fork_handle_stop(struct comp_dev *dev)
+{
+	switch (dev->state) {
+	case COMP_STATE_ACTIVE:
+		/* from flow: extended for multiple inputs */
+		if (fork_status_count(dev, COMP_STATE_ACTIVE))
+			return PPL_STATUS_PATH_STOP;
+
+		if (fork_status_count(dev, COMP_STATE_PAUSED)) {
+			/* only one active pipeline, but other pipelines are paused */
+			dev->cmd_override = COMP_TRIGGER_PAUSE;
+			dev->state = COMP_STATE_PAUSED;
+			return 0;
+		}
+
+		dev->state = COMP_STATE_PREPARE;
+		return 0;
+	}
+
+	return -EINVAL;
```

Comment:
Can we have a trace_err for all fork errors, this will make integration/debug easier to follow. e.g. cant transition from state X to state Y for ABC